compiler-core/src/exhaustiveness.rs

```diff
@@ -29,208 +75,629 @@
+            // out of a branch's checks. So there's no corresponding runtime check
+            // we can perform for them.
+            Pattern::Discard | Pattern::Variable { .. } | Pattern::Assign { .. } => return None,
+            Pattern::Int { value } => {
+                let value = value.clone();
+                RuntimeCheckKind::Int { value }
+            }
+            Pattern::Float { value } => {
+                let value = value.clone();
+                RuntimeCheckKind::Float { value }
+            }
+            Pattern::String { value } => {
+                let value = value.clone();
+                RuntimeCheckKind::String { value }
+            }
+            Pattern::StringPrefix { prefix, .. } => {
+                let prefix = prefix.clone();
+                RuntimeCheckKind::StringPrefix { prefix }
+            }
+            Pattern::Tuple { elements } => {
+                let size = elements.len();
+                RuntimeCheckKind::Tuple { size }
+            }
+            Pattern::Constructor { variant_index, .. } => {
+                let index = *variant_index;
+                RuntimeCheckKind::Variant { index }
+            }
+            Pattern::BitArray { value } => {
+                let value = value.clone();
+                RuntimeCheckKind::BitArray { value }
+            }
+            Pattern::List { .. } => RuntimeCheckKind::NonEmptyList,
+            Pattern::EmptyList => RuntimeCheckKind::EmptyList,
+        };
+
+        Some(kind)
     }
 }
 
-/// A column in a pattern matching table.
+/// This is one of the checks we can take at runtime to decide how to move
+/// forward in the decision tree.
 ///
-/// A column contains a single variable to test, and a pattern to test against
-/// that variable. A row may contain multiple columns, though this wouldn't be
-/// exposed to the source language (= it's an implementation detail).
-#[derive(Clone, Eq, PartialEq, Debug)]
-pub struct Column {
-    variable: Variable,
-    pattern: PatternId,
+/// After performing a successful check on a value we will discover something
+/// about its shape: it might be an int, an variant of a custom type, ...
+/// Some values (like variants and lists) might hold onto additional data we
+/// will have to pattern match on: in order to do that we need a name to refer
+/// to those new variables we've discovered after performing a check. That's
+/// what `args` is for.
+///
+/// Let's have a look at an example. Imagine we have a pattern like this one:
+/// `a is Wibble(1, _, [])`; after performing a runtime check to make sure `a`
+/// is indeed a `Wibble`, we'll need to perform additional checks on it's
+/// arguments: that pattern will be replaced by three new ones `a0 is 1`,
+/// `a1 is _` and `a2 is []`. Those new variables are the `args`.
+///
+#[derive(Clone, Debug)]
+pub struct RuntimeCheck {
+    kind: RuntimeCheckKind,
+    args: Vec<Variable>,
```

Comment:
Why are the arguments not within the enum? There's a implicit relationship between the number of arguments and the value of the other field, but due to the type design not reflecting that the compiler has to assert the length when using them.